### ISSUE TITLE
feat(mat-button): add progress-spinner option

### ISSUE THREAD

**Issue by CaerusKaru** [angular/material2#13667]
#### Bug, feature request, or proposal:

Proposal

#### What is the expected behavior?

Allow for a properly-sized and positioned progress spinner to be triggered inside a `mat-button`, temporarily overriding its initial content

#### What is the current behavior?

No configuration available, can manually place and position a progress spinner inside a `mat-button`

#### What are the steps to reproduce?

This is a sample template thrown together (very unoptimized): https://stackblitz.com/edit/angular-material2-issue-qbjrmi?file=app%2Fapp.component.html

Click on `LOADING` to trigger the spinner

#### What is the use-case or motivation for changing an existing behavior?

Defined in the spec: https://material.io/design/components/progress-indicators.html#circular-progress-indicators under “Integrating with actions”

#### Is there anything else we should know?

Probably needs an a11y review since the content of the button will change

**Comment by CaerusKaru** [comment]
Proposed API:
* The code would live in `mat-button`, with an added configuration option. This would add a dependency on the `MatProgressSpinnerModule` (though we may be able to do some DI to avoid the bloat)
* Another option would be to have the spinner live elsewhere in the template, and use a `TemplateRef` to link to it in the button (like `mat-menu`
* The option would be for a manual trigger to shift the template to the spinner, like `matProgress="someBoolean"`

**Comment by jelbourn** [comment]
We definitely can't introduce a dependency on spinner from button. Rather than making it specific to button, what about something that more generically adds a spinner to some arbitrary container?

**Comment by CaerusKaru** [comment]
Yeah that's what I was thinking with option 2: we could create a mixin or something for each container component that can accept a `mat-spinner`. I'm just trying to think of other "container" use cases, do you have another one in mind?

**Comment by CaerusKaru** [comment]
After looking at it, there are definitely use cases beyond `mat-button`, for instance `mat-menu` or `mat-select`, where content can be lazy-loaded (or even just respond to user events).

So here's the proposed design:
* We create a new container component called `mat-progress-container` that doesn't render on the page, but can be triggered and injected into various other components:
```html
<mat-progress-container #progress>
  <mat-progress-spinner mode="indeterminate" diameter="25"></mat-progress-spinner>
</mat-progress-container>
<button mat-raised-button matProgress="progress" [matProgressTrigger]="someBoolean">SUBMIT</button>
```
* This would allow us, when we add the `matProgress` directive to transclude anything when `matProgressTrigger` is `false`, and replace it with the `ng-template` from the container when `true`

**Comment by jelbourn** [comment]
I don't quite follow what the `matProgress` directive is doing in that example. 

Also, you are right that changing the content of the button will break the accessibility of the element. I'm not sure if adding a `role="progress"` inside of a button is a supported interaction; we'd have to try it out on the screen-readers we support (especially JAWS) and see how they respond.

**Comment by CaerusKaru** [comment]
It's analogous to the menu API `[matMenuTriggerFor]="menu"`, but instead of triggering an overlay somewhere to appear, it replaces the content of the button with the `TemplateRef` contents.

**Comment by CaerusKaru** [comment]
As for a11y, I was looking at [this](https://codeburst.io/how-to-create-a-simple-css-loading-spinner-make-it-accessible-e5c83c2e464c?gi=e2c15caabe2e) article on doing the same but just with CSS. It seems applicable, though it would have to go through the same review.

**Comment by CaerusKaru** [comment]
Discussed with @crisbeto offline, and we've come up with two options:

* We'll introduce a `mat-progress-container`, but its purpose is to center content inside another container element like `mat-button` or `mat-menu`
* We'll also introduce a `matProgress` directive that sits on the container (ie the `mat-button` or `mat-menu`) to set `position: relative` and necessary ARIA attributes based on a boolean trigger

Here's where the ideas diverge:
1. In the first idea, we leave the rest to the user, likely to use this template:
```html
<button mat-raised-button [matProgress]="trigger">
  <mat-progress-container *ngIf="trigger; else content">...</mat-progress-container>
  <ng-template #content>Upload</ng-template>
</button>
```
2. In the second idea, the burden of templating is on us. IMO this makes for a much nicer UX, but complicates the code a bit
```html
<button mat-raised-button [matProgress]="trigger">
  <mat-progress-container>...</mat-progress-container>
  <mat-progress-content>Upload</mat-progress-content>
</button>
```
This not only provides a clear differentiation between the progress and the content, it also reduces the redundancy and possibility for failure by removing the dual `trigger` uses. Here, the `progress-container` is responsible for marshaling content inside the button/menu, replacing the content based on the value of the host's progress directive.

The only issues with both approaches are a) the best way to handle a11y, and b) the best way to swap content while maintaining size concerns (i.e. making sure the container height stays the same)

**Comment by arman-g** [comment]
Any updates on this one?

**Comment by mcaden** [comment]
Until this is in, there's a fairly simple directive that covers it.

Usage that swaps out an icon:
```html
<button ... [disabled]="isLoading"><mat-icon *appSpinOn="isLoading">save</mat-icon> Submit</button>
```

Or an empty span to not have an icon
```html
<button ... [disabled]="isLoading"><span *appSpinOn="isLoading"></span> Submit</button>
```


Directive:
```ts
/**
 * This directive swaps the attached template for a spinner based upon some expression
 * @example `<mat-icon *appSpinOn="isLoading"></mat-icon>`
 */
@Directive({
  selector: '[appSpinOn]',
})
export class SpinOnDirective {
  private hasView = false;

  constructor(
    private componentFactoryResolver: ComponentFactoryResolver,
    private templateRef: TemplateRef<any>,
    private viewContainer: ViewContainerRef
  ) {}

  @Input() set appSpinOn(condition: boolean) {
    if (!condition && !this.hasView) {
      this.viewContainer.clear();
      this.viewContainer.createEmbeddedView(this.templateRef);
      this.hasView = true;
    } else if (condition && this.hasView) {
      this.viewContainer.clear();
      this.addSpinner();
      this.hasView = false;
    }
  }

  addSpinner() {
    const componentFactory = this.componentFactoryResolver.resolveComponentFactory(MatProgressSpinner);
    this.viewContainer.clear();

    const componentRef = this.viewContainer.createComponent<MatProgressSpinner>(componentFactory);
    componentRef.instance.diameter = 18;
    componentRef.instance.color = 'accent';
    componentRef.instance.mode = 'indeterminate';
    componentRef.instance._elementRef.nativeElement.classList.add('button-spinner');
  }
}
```

CSS:
```css
.button-spinner, .mat-progress-spinner.button-spinner {
  vertical-align: middle;
  display: inline-block;
}
```

**Comment by arknu** [comment]
How come this is still not implemented? Is is specifically stated in the Material Design guidelines that you should "Use circular indicators for short, indeterminate activities (< 5 seconds) requiring user attention." (https://material.io/components/progress-indicators#circular-progress-indicators) - i.e. anything that requires a trip to the server when clicking the button. If this library is supposed to follow the Material Design guidelines, this feature needs to be implemented.
I know you can do it manually, but this really is something that I would expect a library like Angular Material to just take care of. 
Of the proposed APIs by @CaerusKaru, I'd definitely prefer option 2. I don't think you should have to manually do *ngIf and templates to make this feature work.

**Comment by xdHampus** [comment]
:skull:

**Comment by jermowery** [comment]
I am going to take a look at this, but I cannot promise anything. There is a lot of complexity with this:

- Performance (can this be done without a structural directive?)
- Contrast (what spinner colors makes sense when inside of a button?)
- Focus management
- Disable states
- Hiding the text in a way that makes sense for screen readers

This is by no means a simple problem to solve, but over the coming weeks I will see if I can manage something

**Comment by jermowery** [comment]
Discussed this a bit with @andrewseguin a key design goal is avoiding requiring downloading the JS for the progress spinner when any button is used, only when a button with a progress spinner is used.

Back in 2018 this meant doing something like the approach above, but in 2025 we have `@defer`, so it might be possible make the API for this just a boolean in the button component

**Comment by jermowery** [comment]
I threw together a simple version of doing this with `@defer` internally and it appears to work super well. The behaviors all look good, still some technical stuff that would need to be worked out but it seems promising

**Comment by mcaden** [comment]
Exciting! I've been away from Angular day-to-day for a few years but I'm very interested in what an optimized version could look like now.

**Comment by jermowery** [comment]
Tried my best at this but I am also abandoning my efforts here. 

Using `@defer` won't work for apps that can't handle dynamic imports. The apps I am making are using a serving system that can handle this but it isn't applicable more broadly so this isn't viable. 

Instead I am making a directive (non-structural) which adjusts the DOM of the MatButton to add a spinner container element and then adds the spinner to this container when requested to do so. 

In theory, MatButton could have an API that does something similar, content project a mat spinner into a container and then only show that container when an input says to do so.

**Comment by jermowery** [comment]
After having some internal discussions, I started implementing an approach to this in https://github.com/jermowery/components/pull/1

There are still many open questions though
